### ISSUE TITLE
Examples: OB 2.0 Assertion issued to a DID that may also be a valid Verifiable Credentials

### ISSUE THREAD

**Issue by ottonomy** [IMSGlobal/openbadges-specification#267]
Here is an assertion that has the structure of a valid OB assertion that also is in the structure of a valid Verifiable Credentials 

An example of using a DID as a recipient identifier property in an assertion:

```jsonld
{
  "@context": [
    "https://w3id.org/openbadges/v2"
  ],
  "id": "https://example.com/assertions/1001",
  "type": [
    "Assertion"
  ],

  "recipient": {
    "type": "id",
    "identity": "did:web:recipient.example.com"
  },

  "issuedOn": "2020-03-10T04:24:12.164Z",
  "badge": "https://example.com/badgeclasses/123",
  "verification": {
    "type": "hosted"
  }
}
```

The `recipient.type` property value `"id"` is not in the allowed list of values for that property but it is in the context like the other properties that are identified as [Profile Identifier Properties](https://openbadgespec.org#ProfileIdentifierProperties). This would fail validation. We need an errata update to the 2.0 spec to allow it and then a validator update to accept it. If we wanted to do the above change, it would be a very minor update to 2.0 that would not disrupt existing use cases. 

Alternatively, if we wanted to allow DID-based recipient identification without making this update, there has been some discussion of using an extension. However, an extension cannot modify the allowed values of core properties (the extension schema might pass but the badge would not, so the extension data would need to be separate from present valid recipient data. In addition, the `recipient.type` field cannot be modified to be an array that includes `"Extension"`, so the extension data would need to sit outside of the `recipient` property. Here is an example showing how an extension could be used to supplement a valid 2.0 assertion with DID-based recipient identification. But it would still require awarding the badge to a currently-supported identifier like email or url and only clients that knew specifically to look for the extension would be able to use the DID data.

```jsonld
{
  "@context": [
    "https://w3id.org/openbadges/v2"
  ],
  "id": "https://example.com/assertions/1001",
  "type": [
    "Assertion"
  ],

  "recipient": {
    "type": "email",
    "identity": "alice@example.com"
  },
  "https://exampleextensions.org/recipientDid": {
    "type": ["Extension", "https://exampleextensions.org/recipientDid/#RecipientDidExtension"],
    "@context": "https://exampleextensions.org/recipientDid/context.json",
    "didIdentity": "did:web:recipient.example.com""
  },

  "issuedOn": "2020-03-10T04:24:12.164Z",
  "badge": "https://example.com/badgeclasses/123",
  "verification": {
    "type": "hosted"
  }
}
```

Here's an additional example showing how properties from Verifiable Credentials Data Model could be added to an assertion without disturbing the core OB properties. This could be done without the above suggested errata change, so that the VC is awarded to a DID, but the Assertion is awarded to an email. It could be upgraded to include DID-based identification for the Open Badges properties if the errata change is accepted and published.
```jsonld
{
  "@context": [
    "https://www.w3.org/2018/credentials/v1",
    "https://w3id.org/openbadges/v2"
  ],
  "id": "https://example.com/assertions/1001",
  "type": [
    "VerifiableCredential",
    "Assertion"
  ],
  
  "cred:issuer": {
    "id": "did:web:edx.com"
  },
  
  "issuanceDate": "2020-03-10T04:24:12.164Z",
  "issuedOn": "2020-03-10T04:24:12.164Z",

  "recipient": {
    "type": "email",
    "identity": "alice@example.com"
  },
  "credentialSubject": {
    "id": "did:web:recipient.example.com",
    "schema:hasAchieved": {
      "id": "https://example.com/badgeclasses/123",
      "type": "BadgeClass",
      "name": "Robotics Badge",
      "image": "data:image/png;base64,...",
      "description": "Awarded to people who have built a robot",
      "criteria": {
        "narrative": "Build a Robot and post a picture of it"
      },
      "issuer": "https://example.com/issuer"
    }
  },
  "badge": "https://example.com/badgeclasses/123",
  
  "proof": {
    "type": "PROOF-Type",
    "created": "DATE OF SIGNING",
    "proofPurpose": "assertionMethod",
    "verificationMethod": "ISSUER-SIGNING-KEY-DID-LINK",
    "jws": "SIGNATURE"
  },
  "verification": {
    "type": "hosted"
  }
}
```

Then the above approach including also some Verifiable Credentials Data Model information in collaboration with the communities that use that.

**Comment by ViktorHaag** [comment]
On the whole, I favour the first approach that would require a minor enhancement of the existing spec, owing to its simplicity.

**Comment by kayaelle** [comment]
It's great to see this. I agree with @ViktorHaag. 

A couple of riffs:

- what if the recipient type was more specific and instead of "id", "DID" was used? Email, URL, & Telephone # have anticipated formats. ID doesn't specify what type of id and some people may use that for anything, making it challenging to use it for DIDs specifically.
- Also, would this be a good time to make it possible to have more than one profile identifiers in a badge and possibly more than one recipient?

**Comment by glcohen** [comment]
@kayaelle my initial proposal did just what you described. the point was raised that for LD-tie in using the ID property might be nicer.

@ottonomy this is great! thanks for putting it together. While it works, I'm hesitant to push for this implementation as "the example" to add DIDs to Badges. Reason being, it brings in all the complexity of, and essentially merges the VC and Badge spec. 

I think it should be _an option_, but the one we push should be far simpler, just including DIDs.

**Comment by ottonomy** [comment]
@kayaelle yep I recognized that you recommended the same approach that @glcohen recommeded previously. I'm not super-strong on this recommendation, especially seeing other experts keep recommending the same alternative. I used `"id"` here because the other options for this property, like `"email"` and `"telephone"` are mapped in the V2 context to properties that would appear on the user profile. An IRI is required for this value (though we could map "did" to an IRI if needed) because the `"type"` field is aliased to `"@type"`. If a DID appears on a recipient profile as a property, the property it appears under is likely `"id"`, not `"did"`. This is not a very strong reason to go with this approach, but maybe a good enough weak reason. Is there an overriding reason to use `"type": "did"` ?

@glcohen Yeah, it's good to be able to solve only one question at once. In the first example in the initial post is just a DIDs solution in OB with no VC structure. Some implementers in the VC community are likely to pick up the rest of this work in their own implementations, and there is steam gathering around the idea of using a `schema:hasAchieved` claim, which is a term that has been proposed to complement the existing EducationalOccupationalCredential work in schema.org.

**Comment by kayaelle** [comment]
I'm taking a fresh look at this.
